File: event-service/src/main/java/com/github/manosbatsis/eventservice/mapper/UserMapper.java
package com.github.manosbatsis.eventservice.mapper;

import com.github.manosbatsis.eventservice.model.UserEvent;
import com.github.manosbatsis.eventservice.model.UserEventKey;
import com.github.manosbatsis.eventservice.rest.dto.UserEventResponse;
import com.github.manosbatsis.userservice.messages.UserEventMessage;
import java.util.Date;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.springframework.context.annotation.Configuration;
import org.springframework.messaging.Message;

@Configuration
@Mapper(componentModel = "spring")
public interface UserMapper {

  @Mapping(source = "key.userId", target = "userId")
  @Mapping(source = "key.datetime", target = "datetime")
  UserEventResponse toUserEventResponse(UserEvent userEvent);

  default UserEvent createUserEvent(Message<UserEventMessage> message) {
    UserEventMessage userEventMessage = message.getPayload();
    UserEvent userEvent = new UserEvent();
    userEvent.setKey(
        new UserEventKey(
            userEventMessage.getUserId(), new Date(userEventMessage.getEventTimestamp())));
    userEvent.setType(userEventMessage.getEventType().toString());
    CharSequence userJson = userEventMessage.getUserJson();
    if (userJson != null) {
      userEvent.setData(userJson.toString());
    }
    return userEvent;
  }
}

File: event-service/src/main/kotlin/com/github/manosbatsis/services/event/mapper/UserMapper.kt
package com.github.manosbatsis.services.event.mapper

import com.github.manosbatsis.lib.core.log.loggerFor
import com.github.manosbatsis.services.event.model.UserEvent
import com.github.manosbatsis.services.event.model.UserEventKey
import com.github.manosbatsis.services.event.rest.dto.UserEventResponse
import com.github.manosbatsis.services.user.messages.UserEventMessage
import java.util.Date

object UserMapper {
    private val log = loggerFor<UserMapper>()
    fun toUserEvent(userEventMessage: UserEventMessage): UserEvent {
        log.warn("toUserEvent, userEventMessage: {}", userEventMessage)
        return UserEvent(
            key = UserEventKey(
                userEventMessage.getUserId(),
                Date(userEventMessage.getEventTimestamp()),
            ),
            type = userEventMessage.getEventType().toString(),
            data = userEventMessage.getUserJson(),
        )
    }

    fun toUserEventResponse(userEvent: UserEvent): UserEventResponse {
        log.warn("toUserEventMessage, userEvent: {}", userEvent)
        return UserEventResponse(
            userEvent.key.userId,
            userEvent.key.datetime,
            userEvent.type,
            userEvent.data,
        )
    }
}
